src/hashset.h

```diff
@@ -0,0 +1,188 @@
+#ifndef HASHSET_H
+#define HASHSET_H
+
+/* Hash table implementation.
+ *
+ * This is a cache-friendly hash table implementation. For details about the
+ * implementation and documentation of functions, see comments in hashset.c.
+ *
+ * The elements in a hashset are of a user-defined type, but an element needs to
+ * contain a key. It can represent a key-value entry, or it can be just a key,
+ * if set semantics are desired.
+ *
+ * Terminology:
+ *
+ * hashset
+ *         An instance of the data structure.
+ *
+ * element
+ *         An element in the hashset. This may be of the same type as the key,
+ *         or a struct containing a key and other fields.
+ * key
+ *         The part of the element used for looking the element up in the hashset.
+ *         May be the entire element or a struct field within the element.
+ *
+ * type
+ *         A struct containing callbacks, such as hash function, key comparison
+ *         function and how to get the key in an element.
+ */
+
+#include "fmacros.h"
+#include <stddef.h>
+#include <stdint.h>
+#include <unistd.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+typedef struct hashsetBucket hashsetBucket;
+typedef struct hashsetStats hashsetStats;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
+    /* Hash function. Defaults to hashing the bits in the pointer, effectively
+     * treating the pointer as an integer. */
+    uint64_t (*hashFunction)(const void *key);
+    /* Compare function, returns 0 if the keys are equal. Defaults to just
+     * comparing the pointers for equality. */
+    int (*keyCompare)(hashset *s, const void *key1, const void *key2);
+    /* Callback to free an element when it's overwritten or deleted.
+     * Optional. */
+    void (*elementDestructor)(hashset *s, void *element);
+    /* Callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. Both tables are already created. */
```

Comment:
No, it isn't meaningless. It affects the point in time the memory usage changes. Kvstore (currenly using dict, but switching to hashset) is using the rehashingStarted callback and within that callback it checks the size of both tables, to keep track of memory overhead during rehashing. The callback calls this function:

```C
/* Provides the number of buckets in the old and new tables during rehashing.
 * To get the sizes in bytes, multiply by HASHSET_BUCKET_SIZE. This function can
 * only be used when rehashing is in progress, and from the rehashingStarted and
 * rehashingCompleted callbacks. */
void hashsetRehashingInfo(hashset *s, size_t *from_size, size_t *to_size) {
```

Do you have an idea how we can reformulate "Both tables are already created" to convey the same information about when the memory allocation happens, without mentioning table creation?

Reply:
The point is that you don't want to expose details of the internal design.  Maybe you don't have "both" tables, maybe instead, you are performing a realloc of the existing table.  I don't know (and shouldn't have to know).

>within that callback it checks the size of both tables

The internal structure should be opaque.  How would it perform this?

I think the key is to understand what is actually important.  Are we just trying to determine the memory growth that's occurring?  I'm not sure why kvstore would need to keep track of memory overhead in hashset.  Maybe hashset needs an API or metrics so that it can report its own overhead?

Reply:
> The point is that you don't want to expose details of the internal design. Maybe you don't have "both" tables, maybe instead, you are performing a realloc of the existing table. I don't know (and shouldn't have to know).

I agree in principle. The problem is that we already expose this information in INFO or MEMORY STATS (or some command like that, don't remember exactly).

To be able to replace the dict for keys, it needs to provide the same info. Otherwise, we have a regression or at least a breaking change.

I want minimal changes to be able to replace the dict and get this thing delivered. I even designed the API to be as close as the dict as possible, to minimize the diff when integrating it in code like kvstore. The diff is already sufficiently large for the integration PR: roughly 1500 lines changed. We can do API changes later and even remove INFO fields, although that would have to be done in a major version with a majority decision.

Reply:
I removed the part of the comment.